### PR TITLE
AARCH64 - movzbl/shrli -> ubfmr2m6 peephole

### PR DESCRIPTION
**Pull request by swalk-cavium**
This creates an ARMv8 specific Vinstr to represent a commonly used form of the
ubfm/ubfx instruction and the peephole to detect it.  This pattern was seen
in several of the regression tests.  It is generated during the lowering of the callm
instruction.

Before
=====
              0x434009c4  53001c21              uxtb w1, w1
              0x434009c8  53027c21              lsr w1, w1, #2
              0x434009cc  d37df021              lsl x1, x1, #3
              0x434009d0  d297c302              movz x2, #0xbe18 
              0x434009d4  f2a1c322              movk x2, #0xe19, lsl #16 
              0x434009d8  f8626821              ldr x1, [x1, x2]
              0x434009dc  d63f0020              blr x1  

After
=====
              0x40400994  53021c21              ubfx w1, w1, #2, #6
              0x40400998  d37df021              lsl x1, x1, #3
              0x4040099c  d2965e02              movz x2, #0xb2f0 
              0x404009a0  f2a081a2              movk x2, #0x40d, lsl #16 
              0x404009a4  f8626821              ldr x1, [x1, x2]
              0x404009a8  d63f0020              blr x1  

This was seen 169 times in hphp/test/quick/all_type_comparison_test.php and
230 times in hphp/test/zend/good/ext/intl/tests/grapheme.php.

The standard regression tests were run with 6 option sets.  No new regressions seen.

### CODE REVIEW

**Comment by swalk-cavium** [comment]
@mxw - Max, can you review this one?  Thanks.

File: hphp/runtime/vm/jit/vasm-arm.cpp [hunk 1]
```diff
@@ -343,6 +343,7 @@ struct Vgen {
   void emit(const fcvtzs& i) { a->Fcvtzs(X(i.d), D(i.s)); }
   void emit(const mrs& i) { a->Mrs(X(i.r), vixl::SystemRegister(i.s.l())); }
   void emit(const msr& i) { a->Msr(vixl::SystemRegister(i.s.l()), X(i.r)); }
+  void emit(const ubfm2r7ml& i) { a->ubfm(W(i.d), W(i.s), 2, 7); }
```
Comment (cmuellner) [inline hphp/runtime/vm/jit/vasm-arm.cpp]: I would prefer a generic ubfm{immr, imms} instruction over a specialised one.

File: hphp/runtime/vm/jit/vasm-simplify.cpp [hunk 2]
```diff
@@ -524,6 +524,24 @@ bool simplify(Env& env, const pop& inst, Vlabel b, size_t i) {
 
 ///////////////////////////////////////////////////////////////////////////////
 
+bool simplify(Env& env, const movzbl& inst, Vlabel b, size_t i) {
+  // movzbl{s, d}; shrli{2, s, d} --> ubfm2r7ml{s, d}
```
Comment (cmuellner) [inline hphp/runtime/vm/jit/vasm-simplify.cpp]: Would be nice if "movzXY{s, tmp} ; shrYi{N, tmp, d} -> ubfm{r, n, s, d}".

File: hphp/runtime/vm/jit/vasm-simplify.cpp [hunk 3]
```diff
@@ -524,6 +524,24 @@ bool simplify(Env& env, const pop& inst, Vlabel b, size_t i) {
...
+  return if_inst<Vinstr::shrli>(env, b, i + 1, [&](const shrli& sh) {
+    if (!(arch() == Arch::ARM &&
+      sh.s0.l() == 2 &&
+      env.use_counts[inst.d] == 1 &&
+      inst.d == sh.s1)) return false;
+
+  return simplify_impl(env, b, i, [&] (Vout& v) {
+    v << copy{inst.s, inst.d};
```
Comment (cmuellner) [inline hphp/runtime/vm/jit/vasm-simplify.cpp]: why copying first and not `ubfm{inst.s, sh.d}`?

Comment (swalk-cavium) [inline hphp/runtime/vm/jit/vasm-simplify.cpp]: Matching up the Vreg types.  These are cleaned up later.

Comment (mxw) [inline hphp/runtime/vm/jit/vasm-simplify.cpp]: What do you mean by "matching up the Vreg types" here?  Both of these are Vreg32...

**Comment by swalk-cavium** [comment]
@mxw - Wouldn't adding immr/imms fields require adding more Immediates to the new intrinsic?
That would be the first Vinstr with multiple immediates.  If I added a different type, I'd have to update the visitor machinery.  Right?

File: hphp/runtime/vm/jit/vasm-simplify.cpp (hunk 3 above)
Comment (swalk-cavium) [inline hphp/runtime/vm/jit/vasm-simplify.cpp]: struct movzbl { Vreg8 s; Vreg32 d; };
struct shrli { Immed s0; Vreg32 s1, d; VregSF sf; Vflags fl; };

The source in the move is Vreg8, the destination in the shrli is a Vreg32,
The copy uses Vreg.  Isn't that the correct way to match up the types?
Or do I not have to worry about that.

**Comment by cmuellner** [comment]
@swalk-cavium you could use a helper function to calculate the values for immr and imms. I think vixl has such a function.

File: hphp/runtime/vm/jit/vasm-simplify.cpp (hunk 3 above)
Comment (cmuellner) [inline hphp/runtime/vm/jit/vasm-simplify.cpp]: A common solution for that is to use `Vreg`, but I don't know if that is appropriate here.

**Comment by mxw** [comment]
@swalk-cavium—I can't think of what would make vasm-visit fail on instructions with multiple immediate operands.  If it does, then yes, you'll have to fix it... but if it already supports it, then you don't.

File: hphp/runtime/vm/jit/vasm-simplify.cpp (hunk 3 above)
Comment (mxw) [inline hphp/runtime/vm/jit/vasm-simplify.cpp]: Oh, I see.  No, `copy{}` is not the right way to do that—the only way is `movzbl{}`.  You'll have to have your new instruction take a Vreg8 src and a Vreg32 dst operand if that's your intention.

**Comment by swalk-cavium** [comment]
@cmuellner - I'm not sure which vixl helper you're referring to.  Can you give me a reference?
We're crossing a level of abstraction and I chose to include information in the name.  Apparently
that's not the right way.

File: hphp/runtime/vm/jit/vasm-simplify.cpp (hunk 3 above)
Comment (cmuellner) [inline hphp/runtime/vm/jit/vasm-simplify.cpp]: In this case we could make ubfmbi{} ubfmwi{} ubfmli{} ubfmqi{}.

**Comment by cmuellner** [comment]
@swalk-cavium I just checked. There is indeed no helper. But I think it is not needed. `immr = sh.s0.l()` and imms depends on the truncation instruction.

**Comment by edwinsmith** [comment]
Many instructions already have 2+ immediates. As long as you have I(whatever) for each one in the instruction's visitor macro, should be fine.

File: hphp/runtime/vm/jit/vasm-simplify.cpp (hunk 3 above)
Comment (mxw) [inline hphp/runtime/vm/jit/vasm-simplify.cpp]: Agreed; that sounds good to me.

**Comment by swalk-cavium** [comment]
@mxw , @cmuellner - Updated per comments above.  Regression tests with 6 option sets run.  No new issues seen.

File: hphp/runtime/vm/jit/vasm-simplify.cpp [hunk 4]
```diff
@@ -524,6 +524,24 @@ bool simplify(Env& env, const pop& inst, Vlabel b, size_t i) {
...
+  // movzbl{s, d}; shrli{2, s, d} --> ubfmli{2, 7, s, d}
+  return if_inst<Vinstr::shrli>(env, b, i + 1, [&](const shrli& sh) {
+    if (!(arch() == Arch::ARM &&
+      sh.s0.l() == 2 &&
+      env.use_counts[inst.d] == 1 &&
+      inst.d == sh.s1)) return false;
+
+  return simplify_impl(env, b, i, [&] (Vout& v) {
+    v << copy{inst.s, inst.d};
+    v << ubfmli{2, 7, inst.d, sh.d, sh.sf, sh.fl};
```
Comment (mxw) [inline hphp/runtime/vm/jit/vasm-simplify.cpp]: Does this have the same effect on the status flags as the `shrli{}`?

File: hphp/runtime/vm/jit/vasm-simplify.cpp [hunk 5]
```diff
@@ -524,6 +524,24 @@ bool simplify(Env& env, const pop& inst, Vlabel b, size_t i) {
...
 
+bool simplify(Env& env, const movzbl& inst, Vlabel b, size_t i) {
+  // movzbl{s, d}; shrli{2, s, d} --> ubfmli{2, 7, s, d}
+  return if_inst<Vinstr::shrli>(env, b, i + 1, [&](const shrli& sh) {
+    if (!(arch() == Arch::ARM &&
+      sh.s0.l() == 2 &&
+      env.use_counts[inst.d] == 1 &&
+      inst.d == sh.s1)) return false;
+
+  return simplify_impl(env, b, i, [&] (Vout& v) {
```
Comment (mxw) [inline hphp/runtime/vm/jit/vasm-simplify.cpp]: Please fix the indentation here and below.

File: hphp/runtime/vm/jit/vasm-simplify.cpp (hunk 4 above)
Comment (swalk-cavium) [inline hphp/runtime/vm/jit/vasm-simplify.cpp]: @mxw - That's a good question.  I'm not seeing anywhere that the flag is being set/used
by either of the shrli/shrqi instructions.  The underlying ARM instruction doesn't touch any
flags.  I'll check further, but it's probably safe to remove from ubfmli.

**Pull request closed without merging**